zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip1559.rs

```diff
@@ -0,0 +1,152 @@
+use crate::{
+    evm_circuit::{
+        util::{
+            constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+            math_gadget::{AddWordsGadget, IsEqualGadget, LtWordGadget, MulWordByU64Gadget},
+            or, Expr, Word,
+        },
+        witness::Transaction,
+    },
+    table::TxFieldTag,
+};
+use eth_types::{geth_types::TxType, Field, ToLittleEndian, U256};
+use halo2_proofs::plonk::{Error, Expression};
+
+/// Transaction EIP-1559 gadget to check sender balance before transfer
+#[derive(Clone, Debug)]
+pub(crate) struct TxEip1559Gadget<F> {
+    is_eip1559_tx: IsEqualGadget<F>,
+    // MaxFeePerGas
+    gas_fee_cap: Word<F>,
+    // MaxPriorityFeePerGas
+    gas_tip_cap: Word<F>,
+    mul_gas_fee_cap_by_gas: MulWordByU64Gadget<F>,
+    balance_check: AddWordsGadget<F, 3, true>,
+    // Error condition
+    // <https://github.com/ethereum/go-ethereum/blob/master/core/state_transition.go#L255>
+    is_insufficient_balance: LtWordGadget<F>,
+    // Error condition
+    // <https://github.com/ethereum/go-ethereum/blob/master/core/state_transition.go#L304>
+    gas_fee_cap_lt_gas_tip_cap: LtWordGadget<F>,
+}
+
+impl<F: Field> TxEip1559Gadget<F> {
+    pub(crate) fn construct(
+        cb: &mut EVMConstraintBuilder<F>,
+        tx_id: Expression<F>,
+        tx_type: Expression<F>,
+        tx_gas: Expression<F>,
+        tx_l1_fee: &Word<F>,
+        value: &Word<F>,
+        sender_balance: &Word<F>,
+    ) -> Self {
+        let is_eip1559_tx = IsEqualGadget::construct(cb, tx_type, (TxType::Eip1559 as u64).expr());
+
+        let [gas_fee_cap, gas_tip_cap] =
+            [TxFieldTag::MaxFeePerGas, TxFieldTag::MaxPriorityFeePerGas]
+                .map(|field_tag| cb.tx_context_as_word(tx_id.expr(), field_tag, None));
+
+        let (
+            mul_gas_fee_cap_by_gas,
+            balance_check,
+            is_insufficient_balance,
+            gas_fee_cap_lt_gas_tip_cap,
+        ) = cb.condition(is_eip1559_tx.expr(), |cb| {
+            let mul_gas_fee_cap_by_gas =
+                MulWordByU64Gadget::construct(cb, gas_fee_cap.clone(), tx_gas);
+
+            let min_balance = cb.query_word_rlc();
+            let balance_check = AddWordsGadget::construct(
+                cb,
+                [
+                    mul_gas_fee_cap_by_gas.product().clone(),
+                    value.clone(),
+                    tx_l1_fee.clone(),
+                ],
+                min_balance.clone(),
+            );
+
+            let is_insufficient_balance = LtWordGadget::construct(cb, sender_balance, &min_balance);
+            let gas_fee_cap_lt_gas_tip_cap =
+                LtWordGadget::construct(cb, &gas_fee_cap, &gas_tip_cap);
+
+            cb.require_zero(
+                "Sender balance must be sufficient, and gas_fee_cap >= gas_tip_cap",
+                or::expr([
+                    is_insufficient_balance.expr(),
+                    gas_fee_cap_lt_gas_tip_cap.expr(),
+                ]),
```

Comment:
Because we know that the `LtGadget`'s `expr` gives us a boolean expression, we can use `sum` instead of `or` to simplify the expression (`or` has a higher degree because its implemented as `not(and(a, b))`).

Note that this only works if both operands are known to be boolean and we wish to constrain to `0`.

Reply:
Fix in commit [2fca5e1](https://github.com/scroll-tech/zkevm-circuits/pull/1066/commits/2fca5e18c979e2a5c86e4da144e2677d2c174a99).